core/tracker/incldelay.go

```diff
@@ -0,0 +1,107 @@
+// Copyright © 2022 Obol Labs Inc.
+//
+// This program is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation, either version 3 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of  MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package tracker
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/obolnetwork/charon/app/errors"
+	"github.com/obolnetwork/charon/app/eth2wrap"
+	"github.com/obolnetwork/charon/core"
+)
+
+// epochLag is the number of epochs to lag when calculating inclusion delay (to ensure finality).
+const epochLag = 1
+
+// dutiesFunc returns the duty definitions for a given duty.
+type dutiesFunc func(context.Context, core.Duty) (core.DutyDefinitionSet, error)
+
+// NewInclDelayFunc returns a function that calculates this cluster's attestation inclusion delay for a block.
+//
+// Inclusion delay is the average of the distance between the slot a validator’s attestation
+// is expected by the network and the slot the attestation is actually included on-chain.
+// See https://rated.gitbook.io/rated-documentation/rating-methodologies/ethereum-beacon-chain/network-explorer-definitions/top-screener#inclusion-delay.
+func NewInclDelayFunc(eth2Cl eth2wrap.Client, dutiesFunc dutiesFunc) func(context.Context, core.Slot) error {
+	return newInclDelayFunc(eth2Cl, dutiesFunc, func(delays []int64) {
```

Comment:
nit: this callback function can be written outside as a separate function for better readability